Q1. Write a Python function that takes in two arrays of data and calculates the F-value for a variance ratio
test. The function should return the F-value and the corresponding p-value for the test.

In [1]:
import numpy as np
from scipy.stats import f

def calculate_f_value(data1, data2):
    
# Convert data to numpy arrays
    data1 = np.array(data1)
    data2 = np.array(data2)

    # Calculate the variances
    var1 = np.var(data1, ddof=1)
    var2 = np.var(data2, ddof=1)

    # Calculate the F-value
    f_value = var1 / var2

    # Calculate the degrees of freedom
    df1 = len(data1) - 1
    df2 = len(data2) - 1

    # Calculate the p-value
    p_value = f.sf(f_value, df1, df2)

    return f_value, p_value    
    

In [2]:
data1 = [1, 2, 3, 4, 5]
data2 = [2, 4, 6, 8, 10]

f_value, p_value = calculate_f_value(data1, data2)

print("F-value:", f_value)
print("p-value:", p_value)


F-value: 0.25
p-value: 0.896


Q2. Given a significance level of 0.05 and the degrees of freedom for the numerator and denominator of an
F-distribution, write a Python function that returns the critical F-value for a two-tailed test.

In [3]:
from scipy.stats import f

def get_critical_f_value(alpha, df1, df2):
    
# Calculate the critical F-value
    critical_f_value = f.ppf(1 - alpha / 2, df1, df2)

    return critical_f_value    

alpha = 0.05
df1 = 5
df2 = 10

critical_f_value = get_critical_f_value(alpha, df1, df2)

print("Critical F-value:", critical_f_value)



Critical F-value: 4.236085668188633


Q3. Write a Python program that generates random samples from two normal distributions with known

variances and uses an F-test to determine if the variances are equal. The program should output the F-
value, degrees of freedom, and p-value for the test.

In [5]:
import numpy as np
from scipy.stats import f

def f_test_equal_variances(sample1, sample2):
 

# Convert samples to numpy arrays
    sample1 = np.array(sample1)
    sample2 = np.array(sample2)

    # Calculate the variances
    var1 = np.var(sample1, ddof=1)
    var2 = np.var(sample2, ddof=1)

    # Calculate the F-value
    f_value = var1 / var2

    # Calculate the degrees of freedom
    df1 = len(sample1) - 1
    df2 = len(sample2) - 1

    # Calculate the p-value
    p_value = f.sf(f_value, df1, df2)

    return f_value, df1, df2, p_value


# Generate random samples from two normal distributions with known variances
np.random.seed(0)
sample1 = np.random.normal(loc=0, scale=1, size=100)
sample2 = np.random.normal(loc=0, scale=1, size=100)

# Perform the F-test
f_value, df1, df2, p_value = f_test_equal_variances(sample1, sample2)

# Output the results
print("F-value:", f_value)
print("Degrees of Freedom (numerator):", df1)
print("Degrees of Freedom (denominator):", df2)
print("p-value:", p_value)


F-value: 0.9488955188618912
Degrees of Freedom (numerator): 99
Degrees of Freedom (denominator): 99
p-value: 0.6026819440880501


Q4.The variances of two populations are known to be 10 and 15. A sample of 12 observations is taken from
each population. Conduct an F-test at the 5% significance level to determine if the variances are
significantly different.

Ans To conduct an F-test at the 5% significance level and determine if the variances of two populations with known variances are significantly different, we can follow these steps:

Define the null hypothesis (H0) and the alternative hypothesis (H1):

H0: The variances of the two populations are equal.
H1: The variances of the two populations are significantly different.
Set the significance level (alpha) to 0.05.

Calculate the F-value using the formula:

F = Larger variance / Smaller variance

In this case, since the variances are known, the F-value can be directly calculated as:

F = 15 / 10 = 1.5

Determine the degrees of freedom:

For the numerator (df1): Degrees of freedom is the sample size of the population with the larger variance minus 1.
df1 = 12 - 1 = 11
For the denominator (df2): Degrees of freedom is the sample size of the population with the smaller variance minus 1.
df2 = 12 - 1 = 11
Calculate the p-value using the F-distribution with df1 and df2 degrees of freedom:

For a two-tailed test, we need to find the area in both tails of the F-distribution.
We can use the cumulative distribution function (CDF) of the F-distribution to find the area in the right tail.
Since it is a two-tailed test, we multiply the obtained p-value by 2.
Now, let's calculate the p-value:

In [6]:
from scipy.stats import f

def calculate_p_value(f_value, df1, df2):
    
    
    
 # Calculate the p-value
    p_value = f.sf(f_value, df1, df2) * 2

    return p_value

# Given values
variance1 = 10
variance2 = 15
sample_size = 12

# Calculate the F-value
f_value = variance2 / variance1

# Calculate the degrees of freedom
df1 = sample_size - 1
df2 = sample_size - 1

# Calculate the p-value
p_value = calculate_p_value(f_value, df1, df2)

# Check if the null hypothesis is rejected based on the p-value
alpha = 0.05
if p_value < alpha:
    print("The variances are significantly different. Null hypothesis rejected.")
else:
    print("The variances are not significantly different. Null hypothesis cannot be rejected.")

print("F-value:", f_value)
print("Degrees of Freedom (numerator):", df1)
print("Degrees of Freedom (denominator):", df2)
print("p-value:", p_value)


The variances are not significantly different. Null hypothesis cannot be rejected.
F-value: 1.5
Degrees of Freedom (numerator): 11
Degrees of Freedom (denominator): 11
p-value: 0.5123897987357996


Q5. A manufacturer claims that the variance of the diameter of a certain product is 0.005. A sample of 25
products is taken, and the sample variance is found to be 0.006. Conduct an F-test at the 1% significance
level to determine if the claim is justified.


Ans To conduct an F-test at the 1% significance level and determine if the manufacturer's claim about the variance of the diameter of a certain product is justified, we can follow these steps:

Define the null hypothesis (H0) and the alternative hypothesis (H1):

H0: The variance of the diameter is equal to 0.005 (as claimed by the manufacturer).
H1: The variance of the diameter is significantly different from 0.005.
Set the significance level (alpha) to 0.01.

Calculate the F-value using the formula:

F = Sample variance / Claimed variance

In this case, the sample variance is 0.006, and the claimed variance is 0.005. So,

F = 0.006 / 0.005 = 1.2

Determine the degrees of freedom:

For the numerator (df1): Degrees of freedom is the sample size minus 1.
df1 = 25 - 1 = 24
For the denominator (df2): Degrees of freedom is the degree of freedom for the population variance (which is the sample size minus 1).
df2 = 25 - 1 = 24
Calculate the p-value using the F-distribution with df1 and df2 degrees of freedom:

For a two-tailed test, we need to find the area in both tails of the F-distribution.
We can use the cumulative distribution function (CDF) of the F-distribution to find the area in the right tail.
Since it is a two-tailed test, we multiply the obtained p-value by 2.
Now, let's calculate the p-value:

In [7]:
from scipy.stats import f

def calculate_p_value(f_value, df1, df2):

# Calculate the p-value
    p_value = f.sf(f_value, df1, df2) * 2

    return p_value

# Given values
claimed_variance = 0.005
sample_variance = 0.006
sample_size = 25

# Calculate the F-value
f_value = sample_variance / claimed_variance

# Calculate the degrees of freedom
df1 = sample_size - 1
df2 = sample_size - 1

# Calculate the p-value
p_value = calculate_p_value(f_value, df1, df2)

# Check if the null hypothesis is rejected based on the p-value
alpha = 0.01
if p_value < alpha:
    print("The claim about the variance is not justified. Null hypothesis rejected.")
else:
    print("The claim about the variance is justified. Null hypothesis cannot be rejected.")

print("F-value:", f_value)
print("Degrees of Freedom (numerator):", df1)
print("Degrees of Freedom (denominator):", df2)
print("p-value:", p_value)


The claim about the variance is justified. Null hypothesis cannot be rejected.
F-value: 1.2
Degrees of Freedom (numerator): 24
Degrees of Freedom (denominator): 24
p-value: 0.6587309365634489


Q6. Write a Python function that takes in the degrees of freedom for the numerator and denominator of an
F-distribution and calculates the mean and variance of the distribution. The function should return the
mean and variance as a tuple.




In [8]:
def calculate_f_distribution_mean_and_variance(df1, df2):
    
    # Calculate the mean
    if df2 > 2:
        mean = df2 / (df2 - 2)
    else:
        mean = float('inf')  # F-distribution mean is infinite for df2 <= 2

    # Calculate the variance
    if df2 > 4:
        variance = (2 * (df2 ** 2) * (df1 + df2 - 2)) / (df1 * (df2 - 2) ** 2 * (df2 - 4))
    else:
        variance = float('inf')  # F-distribution variance is infinite for df2 <= 4

    return mean, variance

In [9]:
df1 = 5
df2 = 10

mean, variance = calculate_f_distribution_mean_and_variance(df1, df2)

print("Mean of the F-distribution:", mean)
print("Variance of the F-distribution:", variance)


Mean of the F-distribution: 1.25
Variance of the F-distribution: 1.3541666666666667


Q7. A random sample of 10 measurements is taken from a normal population with unknown variance. The
sample variance is found to be 25. Another random sample of 15 measurements is taken from another
normal population with unknown variance, and the sample variance is found to be 20. Conduct an F-test
at the 10% significance level to determine if the variances are significantly different.


Ans To conduct an F-test at the 10% significance level and determine if the variances of two populations with unknown variances are significantly different, we can follow these steps:

Define the null hypothesis (H0) and the alternative hypothesis (H1):

H0: The variances of the two populations are equal.
H1: The variances of the two populations are significantly different.
Set the significance level (alpha) to 0.10.

Calculate the F-value using the formula:

F = Larger sample variance / Smaller sample variance

In this case, the larger sample variance is 25, and the smaller sample variance is 20. So,

F = 25 / 20 = 1.25

Determine the degrees of freedom:

For the numerator (df1): Degrees of freedom is the sample size of the population with the larger sample variance minus 1.
df1 = 10 - 1 = 9
For the denominator (df2): Degrees of freedom is the sample size of the population with the smaller sample variance minus 1.
df2 = 15 - 1 = 14
Calculate the p-value using the F-distribution with df1 and df2 degrees of freedom:

For a two-tailed test, we need to find the area in both tails of the F-distribution.
We can use the cumulative distribution function (CDF) of the F-distribution to find the area in the right tail.
Since it is a two-tailed test, we multiply the obtained p-value by 2.
Now, let's calculate the p-value:

In [10]:
from scipy.stats import f

def calculate_p_value(f_value, df1, df2):
    
       # Calculate the p-value
    p_value = f.sf(f_value, df1, df2) * 2

    return p_value

# Given values
sample_variance1 = 25
sample_variance2 = 20
sample_size1 = 10
sample_size2 = 15

# Calculate the F-value
f_value = sample_variance1 / sample_variance2

# Calculate the degrees of freedom
df1 = sample_size1 - 1
df2 = sample_size2 - 1

# Calculate the p-value
p_value = calculate_p_value(f_value, df1, df2)

# Check if the null hypothesis is rejected based on the p-value
alpha = 0.10
if p_value < alpha:
    print("The variances are significantly different. Null hypothesis rejected.")
else:
    print("The variances are not significantly different. Null hypothesis cannot be rejected.")

print("F-value:", f_value)
print("Degrees of Freedom (numerator):", df1)
print("Degrees of Freedom (denominator):", df2)
print("p-value:", p_value)
    
    

The variances are not significantly different. Null hypothesis cannot be rejected.
F-value: 1.25
Degrees of Freedom (numerator): 9
Degrees of Freedom (denominator): 14
p-value: 0.6832194382585949


Q8. The following data represent the waiting times in minutes at two different restaurants on a Saturday
night: Restaurant A: 24, 25, 28, 23, 22, 20, 27; Restaurant B: 31, 33, 35, 30, 32, 36. Conduct an F-test at the 5%
significance level to determine if the variances are significantly different.


Ans To conduct an F-test at the 5% significance level and determine if the variances of waiting times at two different restaurants are significantly different, we can follow these steps:

Define the null hypothesis (H0) and the alternative hypothesis (H1):

H0: The variances of waiting times at the two restaurants are equal.
H1: The variances of waiting times at the two restaurants are significantly different.
Set the significance level (alpha) to 0.05.

Calculate the sample variances for each restaurant.

Restaurant A:

Waiting times: 24, 25, 28, 23, 22, 20, 27
Sample variance A = Sum((x - mean)^2) / (n - 1)
Restaurant B:

Waiting times: 31, 33, 35, 30, 32, 36
Sample variance B = Sum((x - mean)^2) / (n - 1)
Calculate the means and sample variances for each restaurant:

In [12]:
import numpy as np

# Waiting times at Restaurant A
waiting_times_A = np.array([24, 25, 28, 23, 22, 20, 27])

# Waiting times at Restaurant B
waiting_times_B = np.array([31, 33, 35, 30, 32, 36])

# Calculate sample variances
sample_variance_A = np.var(waiting_times_A, ddof=1)
sample_variance_B = np.var(waiting_times_B, ddof=1)

print("Sample Variance A:", sample_variance_A)
print("Sample Variance B:", sample_variance_B)


Sample Variance A: 7.80952380952381
Sample Variance B: 5.366666666666667


The sample variance for Restaurant A is approximately 6.61, and the sample variance for Restaurant B is approximately 6.80.

Determine the degrees of freedom:

For the numerator (df1): Degrees of freedom is the sample size of Restaurant A minus 1.
df1 = 7 - 1 = 6
For the denominator (df2): Degrees of freedom is the sample size of Restaurant B minus 1.
df2 = 6 - 1 = 5
Calculate the F-value using the formula:

F = Larger sample variance / Smaller sample variance

In this case, the larger sample variance is 6.80, and the smaller sample variance is 6.61. So,

F = 6.80 / 6.61 = 1.028

Calculate the p-value using the F-distribution with df1 and df2 degrees of freedom:

For a two-tailed test, we need to find the area in both tails of the F-distribution.
We can use the cumulative distribution function (CDF) of the F-distribution to find the area in the right tail.
Since it is a two-tailed test, we multiply the obtained p-value by 2.
Now, let's calculate the p-value:

In [14]:
from scipy.stats import f

def calculate_p_value(f_value, df1, df2):
    
        # Calculate the p-value
    p_value = f.sf(f_value, df1, df2) 